.emacs.d/config.org


#+title: My Emacs Configuration
#+Startup: content
* Basic UI Configuration
#+begin_src emacs-lisp
  (setq inhibit-startup-message t) ;; Disable Message

  (scroll-bar-mode -1) ;;Disable visible scrollbar

  (menu-bar-mode -1) ;;Disable the menu bar

  (setq visible-bell -1) ;;Set up the visible bell

  (show-paren-mode 1) ;; Enable ShowParent

  (tool-bar-mode -1) ;;Disable the toolbar

  (tooltip-mode -1) ;;Disable tooltips

  (column-number-mode) ;;:Enable number
  (global-display-line-numbers-mode t)

  (dolist (mode '(org-mode-hook
                  term-mode-hook
                  shell-mode-hook
                  eshell-mode-hook))
    (add-hook mode (lambda () (display-line-numbers-mode 0))))
#+end_src
** Browser
#+begin_src emacs-lisp
  (setq browse-url-generic-program 
        (executable-find "firefox-bin")
        browse-url-browser-function 'browse-url-generic)
#+end_src
** Keyboard
#+begin_src emacs-lisp
  (global-set-key (kbd "<escape>") 'keyboard-escape-quit)


  (setq mouse-drag-and-drop-region-cross-program t
        dired-mouse-drag-files t)
#+end_src
** Font Configuration
#+begin_src emacs-lisp

  ;;(set-face-attribute 'default nil :font "Fira Code Retina" :height 101)

  ;; Set the fixed pitch face
  ;;(set-face-attribute 'fixed-pitch nil :font "Fira Code Retina" :height 101)

  ;; Set the variable pitch face
  (set-face-attribute 'variable-pitch nil :font "hack" :height 101 :weight 'regular)

#+end_src 

* Package [[https://melpa.org/][Link]]
#+begin_src emacs-lisp
  (require 'package)
  (setq package-archives '(("melpa" . "https://melpa.org/packages/")
                           ("org" . "https://orgmode.org/elpa/")
                           ("elpa" . "https://elpa.gnu.org/packages/")))
  (package-initialize)    
  (unless package-archive-contents
    (package-refresh-contents))
#+end_src
** Use-Package [[https://jwiegley.github.io/use-package/installation/][Link]]
#+begin_src emacs-lisp
  (require 'use-package)
  (setq use-package-always-ensure t)
#+end_src
** Automatically Update Packages [[https://github.com/rranelli/auto-package-update.el/tree/master][Link]]
#+begin_src emacs-lisp
            (use-package auto-package-update
              :custom
              (auto-package-update-interval 7)
              (auto-package-update-prompt-before-update t)
              (auto-package-update-hide-results t)
              :config
              (auto-package-update-maybe)
              (auto-package-update-at-time "09:00"))
#+end_src
** Cl-lib
#+begin_src emacs-lisp
  (require 'cl-lib)
#+end_src

* Doom-Themes [[https://github.com/hlissner/emacs-doom-themes/tree/master][Link]]
#+begin_src emacs-lisp

    (use-package doom-themes
      :init
      (load-theme 'doom-one t))

  (use-package doom-modeline
    :init (doom-modeline-mode 1)
    :custom (doom-modeline-height 15))

#+end_src

** All-The-Icons [[https://github.com/domtronn/all-the-icons.el/tree/master][Link]]
#+begin_src emacs-lisp
 (use-package all-the-icons)
#+end_src

** All-The-Icons-ivy-rich [[https://github.com/seagle0128/all-the-icons-ivy-rich/tree/master][Link]]
#+begin_src emacs-lisp
  (use-package all-the-icons-ivy-rich
    :defer 1
    :config(all-the-icons-ivy-rich-mode 1))

#+end_src

** All-The-Icons-Dired [[https://github.com/wyuenho/all-the-icons-dired/tree/master][Link]]
#+begin_src emacs-lisp


  (use-package all-the-icons-dired
    :hook (dired-mode . all-the-icons-dired-mode))

#+end_src

* Org Mode [[https://orgmode.org][Link]]
#+begin_src emacs-lisp

  (defun dw/org-mode-setup ()
    (org-indent-mode)
    ;; (variable-pitch-mode 1)
    (auto-fill-mode 0)
    (visual-line-mode 1)
    (setq evil-auto-indent nil))

  (use-package org
    :hook (org-mode . dw/org-mode-setup)
    :config
    (setq org-ellipsis " ▾"
          org-hide-emphasis-markers t))


  (use-package org-superstar
    :after org
    :hook (org-mode . org-superstar-mode)
    :custom
    (org-superstar-remove-leading-stars t)
    (org-superstar-headline-bullets-list '("◉" "○" "●" "○" "●" "○" "●")))



  (with-eval-after-load 'org-faces
    (set-face-attribute 'org-document-title nil :font "Cantarell" :weight 'bold :height 1.3)
    (dolist (face '((org-level-1 . 1.2)
                    (org-level-2 . 1.1)
                    (org-level-3 . 1.05)
                    (org-level-4 . 1.0)
                    (org-level-5 . 1.1)
                    (org-level-6 . 1.1)
                    (org-level-7 . 1.1)
                    (org-level-8 . 1.1)))
      (set-face-attribute (car face) nil :font "Cantarell" :weight 'regular :height (cdr face)))


    (set-face-attribute 'org-block nil    :foreground nil :inherit 'fixed-pitch)
    (set-face-attribute 'org-table nil    :inherit 'fixed-pitch)
    (set-face-attribute 'org-formula nil  :inherit 'fixed-pitch)
    (set-face-attribute 'org-code nil     :inherit '(shadow fixed-pitch))
    (set-face-attribute 'org-table nil    :inherit '(shadow fixed-pitch))
    (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
    (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
    (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
    (set-face-attribute 'org-checkbox nil  :inherit 'fixed-pitch))


  (defun dw/org-mode-visual-fill ()
    (setq visual-fill-column-width 100
          visual-fill-column-center-text t)
    (visual-fill-column-mode 1))

  (use-package visual-fill-column
    :hook (org-mode . dw/org-mode-visual-fill))
#+end_src
* Ivy-Mode [[https://github.com/abo-abo/swiper/tree/master][Link]]
#+begin_src emacs-lisp
  (use-package ivy
    :defer 5
    :diminish ivy-mode
    :bind(("C-s" . swiper)
          :map ivy-minibuffer-map
          ("TAB" . ivy-alt-done)
          ("C-f" . ivy-alt-done)
          ("C-l" . ivy-alt-done)
          ("C-j" . ivy-next-line)
          ("C-k" . ivy-previous-line)
          :map ivy-switch-buffer-map
          ("C-k" . ivy-previous-line)
          ("C-l" . ivy-done)
          ("C-d" . ivy-switch-buffer-kill)
          :map ivy-reverse-i-search-map
          ("C-k" . ivy-previous-line)
          ("C-d" . ivy-reverse-i-search-kill))
    :config
    (ivy-mode 1))

  (use-package which-key ;;
    :defer 5
    :diminish which-key-mode
    :config (which-key-mode
             (setq which-key-idle-delay 2)))

  (use-package ivy-rich
    :defer 7
    :config
    (ivy-rich-mode 1))


#+end_src
** Counsel [[https://github.com/abo-abo/swiper/tree/master][Link]]
#+begin_src emacs-lisp

  (use-package counsel
    :bind (("M-x" . counsel-M-x)
           ("C-x b" . counsel-ibuffer)
           ("C-x C-f" . counsel-find-file)
           :map minibuffer-local-map
           ("C-r" . 'counsel-minibuffer-history))
    :config
    (setq ivy-initial-inputs-alist nil)) ;; Don't start searches with ^

#+end_src
* Web
** Astro 
#+begin_src emacs-lisp
(define-derived-mode astro-ts-mode web-mode "astro-ts")
(setq auto-mode-alist
      (append '(("\\.astro\\'" . astro-ts-mode))
              auto-mode-alist))
#+end_src 
***** Eglot
#+begin_src emacs-lisp
  (use-package eglot
    :ensure t
    :config
    ;; Añadir configuración para astro-ts-mode
    (add-to-list 'eglot-server-programs
                 '(astro-ts-mode . ("astro-ls" "--stdio"
                                    :initializationOptions
                                    (:typescript (:tsdk "./node_modules/typescript/lib")))))
    ;; Configurar eglot para HTML
    (add-to-list 'eglot-server-programs
                 '(html-mode . ("/usr/bin/vscode-html-language-server" "--stdio")))

    ;; Si no quieres que eglot se inicie automáticamente con css-mode, scss-mode y less-mode, comenta o elimina las siguientes líneas
    (add-to-list 'eglot-server-programs
                 '((css-mode scss-mode less-mode) . ("/usr/bin/vscode-css-language-server" "--stdio")))

    :init
    ;; auto start eglot para astro-ts-mode
    (add-hook 'astro-ts-mode-hook 'eglot-ensure)

    ;; auto start eglot para HTML
    (add-hook 'html-mode-hook 'eglot-ensure))
#+end_src
**  Web-Mode [[https://github.com/fxbois/web-mode/tree/master][Link]]
#+begin_src emacs-lisp
  (use-package web-mode
  :config
  ;; Add .astro files to web-mode
  (add-to-list 'auto-mode-alist '("\\.html?\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.astro\\'" . web-mode))
  (progn
    (setq web-mode-engines-alist
          '(("django"  . "\\.html\\'")))
    (setq web-mode-ac-sources-alist
          '(("css" . (ac-source-css-property))
            ("html" . (ac-source-work-in-buffer ac-source-abbrev))))
    (setq web-mode-enable-auto-closing t)
    (setq web-mode-enable-auto-quoting t)
    (setq web-mode-enable-current-column-highlight t)
    (setq web-mode-enable-current-element-highlight t)))
#+end_src
** Emmet [[https://github.com/smihica/emmet-mode/tree/master][Link]]
#+begin_src emacs-lisp
  (use-package emmet-mode
  :defer 5
  :hook
  (sgml-mode . emmet-mode) ;; Auto-start on any markup modes
  (web-mode . emmet-mode)  ;; Auto-start on any markup modes
  (css-mode . emmet-mode)  ;; Enable Emmet's CSS abbreviation
  (astro-ts-mode . emmet-mode)
  (php-mode . emmet-mode)
  (js-mode . emmet-mode)
  (typescript-mode . emmet-mode))
 #+end_src
** Company [[https://github.com/company-mode/company-mode/tree/master][Link]]
#+begin_src emacs-lisp
  (use-package company
    :defer 5
    :hook
    (after-init . global-company-mode))

#+end_src
** Prettier-JS [[https://github.com/prettier/prettier-emacs][Link]]
#+begin_src emacs-lisp
(use-package prettier
  :ensure t
  :hook ((web-mode . prettier-mode)
         (css-mode . prettier-mode)
         (js-mode . prettier-mode)
         (typescript-mode . prettier-mode)
         (php-mode . prettier-mode)))
#+end_src
** Flycheck [[https://github.com/flycheck/flycheck][Link]]
#+begin_src emacs-lisp
  (use-package flycheck
    :ensure t
    :init (global-flycheck-mode))
#+end_src
** Yasnippet
#+begin_src emacs-lisp
  (use-package yasnippet
  :ensure t
  :config
  (yas-global-mode 1))
#+end_src

** Lsp-Ui
#+begin_src emacs-lisp
  (use-package lsp-ui
    :ensure t
    :commands lsp-ui-mode)
#+end_src
** Lsp-Mode
#+begin_src emacs-lisp
  (use-package lsp-mode
  :ensure t
  :commands (lsp lsp-deferred)
  :hook ((web-mode . lsp-deferred)
         (css-mode . lsp-deferred)
         (js-mode . lsp-deferred)
         (typescript-mode . lsp-deferred)
         (php-mode . lsp-deferred))
  :init
  (setq lsp-keymap-prefix "C-c l"))
#+end_src
** JavaScript 
***** Rjsx-Mode  [[https://github.com/felipeochoa/rjsx-mode][Link]]
#+begin_src emacs-lisp
 (use-package rjsx-mode
	   :mode "\\.js\\'")
#+end_src
***** Typescript  [[https://github.com/ananthakumaran/tide][Link]]
#+begin_src emacs-lisp
  (use-package tide
    :ensure t
    :after (rjsx-mode company flycheck)
    :hook ((rjsx-mode . setup-tide-mode)
           (typescript-mode . setup-tide-mode)
           (js-mode . setup-tide-mode)))

  (defun setup-tide-mode ()
    (interactive)
    (tide-setup)
    (flycheck-mode +1)
    (setq flycheck-check-syntax-automatically '(save mode-enabled))
    (eldoc-mode +1)
    (tide-hl-identifier-mode +1)
    (company-mode +1))

  ;; Rjsx Mode para JS
  (use-package rjsx-mode
    :ensure t
    :mode "\\.js\\'")

  ;; Configuración de Typescript Mode
  (use-package typescript-mode
    :ensure t
    :mode "\\.ts\\'")


#+end_src
** Css
#+begin_src emacs-lisp
  (use-package css-mode
    :ensure t
    :mode "\\.css\\'")
#+end_src

** PHP
#+begin_src emacs-lisp
;; Configuración de PHP Mode
(use-package php-mode
  :ensure t
  :mode "\\.php\\'")

#+end_src
* Magit [[https://magit.vc/][Link]]
#+begin_src emacs-lisp

  (use-package magit
    :commands (magit-status magit-get-current-branch)
    :custom
    (magit-display-buffer-function #'magit-display-buffer-same-window-except-diff-v1))

#+end_src

* Projectile [[https://github.com/bbatsov/projectile][Link]]
#+begin_src emacs-lisp
   (use-package projectile
    :demand t
    :config
    (projectile-mode) ; Activar el modo de proyecto
    (projectile-global-mode 1) ; Activar el modo global de projectile
    (setq projectile-project-search-path '(("~/-/" . 4) ("~/Projects/" . 4)))
    (setq projectile-switch-project-action #'projectile-dired)
    :bind (("C-M-p" . projectile-find-file)
           ("C-c p" . projectile-command-map)))

  (use-package helm-projectile
    :ensure t
    :config (helm-projectile-on))


    #+end_src
* Telegram [[https://github.com/zevlg/telega.el][Link]]
#+begin_src emacs-lisp

  (defun dw/after-tracking-remove-buffer (buffer)
    (dw/update-polybar-telegram))

  (advice-add 'tracking-add-buffer :around #'dw/around-tracking-add-buffer)
  (advice-add 'tracking-remove-buffer :after #'dw/after-tracking-remove-buffer)
  (advice-remove 'tracking-remove-buffer #'dw/around-tracking-remove-buffer)

  ;; Advise exwm-workspace-switch so that we can more reliably clear tracking buffers
  ;; NOTE: This is a hack and I hate it.  It'd be great to find a better solution.
  (defun dw/before-exwm-workspace-switch (frame-or-index &optional force)
    (when (fboundp 'tracking-remove-visible-buffers)
      (when (eq exwm-workspace-current-index 0)
        (tracking-remove-visible-buffers))))

  (advice-add 'exwm-workspace-switch :before #'dw/before-exwm-workspace-switch)

  (use-package telega
    :commands telega
    :config
    (setq telega-user-use-avatars nil
          telega-use-tracking-for '(any pin unread)
          telega-chat-use-markdown-formatting t
          telega-emoji-use-images t
          telega-completing-read-function #'ivy-completing-read
          telega-msg-rainbow-title nil
          telega-chat-fill-column 75))


#+end_src
* Dashboard [[https://github.com/emacs-dashboard/emacs-dashboard][Link]]
#+begin_src emacs-lisp
  (use-package dashboard
    :ensure t
    :config
    (dashboard-setup-startup-hook))

  ;; Set the title

  (setq dashboard-banner-logo-title "      Welcome to my Emacs\n Jeet-u. All Rights Reserved. ")
  ;; Set the banner
  (setq dashboard-startup-banner "~/Downloads/milogo.png")
  ;; Value can be:
  ;;  - 'official which displays the official emacs logo.
  ;;  - 'logo which displays an alternative emacs logo.
  ;;  - an integer which displays one of the text banners
  ;;    (see dashboard-banners-directory files).
  ;;  - a string that specifies a path for a custom banner
  ;;    currently supported types are gif/image/text/xbm.
  ;;  - a cons of 2 strings which specifies the path of an image to use
  ;;    and other path of a text file to use if image isn't supported.
  ;;    ("path/to/image/file/image.png" . "path/to/text/file/text.txt").
  ;;  - a list that can display an random banner,
  ;;    supported values are: string (filepath), 'official, 'logo and integers.

  ;; Content is not centered by default. To center, set
  ;; (setq dashboard-center-content t)
  ;; vertically center content
  (setq dashboard-center-content t)
  (setq dashboard-vertically-center-content t)

  ;; To disable shortcut "jump" indicators for each section, set
  (setq dashboard-show-shortcuts nil)


  (setq dashboard-items '((projects   . 5)
                          (recents . 5)
                          (bookmarks  . 5)))






  (setq dashboard-startupify-list '(dashboard-insert-banner
                                    dashboard-insert-newline
                                    dashboard-insert-banner-title
                                    dashboard-insert-newline
                                    dashboard-insert-navigator
                                    dashboard-insert-newline
                                    dashboard-insert-init-info
                                    dashboard-insert-items
                                    dashboard-insert-newline
                                    dashboard-insert-footer))


  (setq dashboard-icon-type 'all-the-icons)

  ;;(setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)

  (setq dashboard-navigator-buttons
        `(
          ;; Botones en la misma línea
          ((,(all-the-icons-octicon "mark-github" :height 1.1 :v-adjust 0.0)
            "Homepage"
            "Browse homepage"
            (lambda (&rest _) (browse-url "https://github.com/Jeet-u")))
           (,(all-the-icons-fileicon "gentoo" :height 1.1 :v-adjust 0.0)
            "Gentoo"
            ""
            (lambda (&rest _) (browse-url "https://wiki.gentoo.org/wiki/Handbook:Main_Page")))
           (,(all-the-icons-faicon "envelope" :height 1.1 :v-adjust 0.0)
            "Proton"
            ""
            (lambda (&rest _) (browse-url "https://protonmail.com")))
           (,(all-the-icons-faicon "firefox" :height 1.1 :v-adjust 0.0)
            "Nyxt"
            ""
            (lambda (&rest _) (start-process "nyxt" nil "/usr/bin/nyxt"))))
          ))


  (require 'bookmark)

#+end_src

* Vterm-toogle [[https://github.com/jixiuf/vterm-toggle?tab=readme-ov-file][Link]]
#+begin_src emacs-lisp

  (use-package vterm
  :ensure t)
  
  (setq vterm-toggle-fullscreen-p nil)
  (add-to-list 'display-buffer-alist
               '((lambda (buffer-or-name _)
                     (let ((buffer (get-buffer buffer-or-name)))
                       (with-current-buffer buffer
                         (or (equal major-mode 'vterm-mode)
                             (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
                  (display-buffer-reuse-window display-buffer-at-bottom)
                  ;;(display-buffer-reuse-window display-buffer-in-direction)
                  ;;display-buffer-in-direction/direction/dedicated is added in emacs27
                  ;;(direction . bottom)
                  ;;(dedicated . t) ;dedicated is supported in emacs27
                  (reusable-frames . visible)
                  (window-height . 0.3)))
#+end_src
* Ace-Window [[https://github.com/abo-abo/ace-window][Link]]
#+begin_src emacs-lisp
  (use-package ace-window
    :bind
    ("C-x o" . ace-window)
    :config
    (setq aw-keys '(?a ?s ?d ?f)))
#+end_src

* Delimiters [[https://github.com/Fanael/rainbow-delimiters][Link]]
#+begin_src emacs-lisp
  (use-package rainbow-delimiters
    :defer 5
    :hook (prog-mode . rainbow-delimiters-mode))
#+end_src

* Keylogs [[https://github.com/lewang/command-log-mode][Link]]
#+begin_src emacs-lisp
(use-package command-log-mode
  :defer t)
#+end_src
* HelpFul [[https://github.com/Wilfred/helpful/tree/master][Link]]
#+begin_src emacs-lisp
  (use-package helpful
    :defer 5
    :custom
    (counsel-describe-function-function #'helpful-callable)
    (counsel-describe-variable-function #'helpful-variable)
    :bind
    ([remap describe-function] . helpful-function)
    ([remap describe-symbol] . helpful-symbol)
    ([remap describe-variable] . helpful-variable)
    ([remap describe-command] . helpful-command)
    ([remap describe-key] . helpful-key))


#+end_src






dwm/ConfigurationDwm.org
#+title: My Dwm Configuration
#+Startup content


* Scripts DWM
** [[https://dwm.suckless.org/patches/taggrid/][Tagrid]]
** [[https://dwm.suckless.org/patches/alpha/][alpha]]
** [[https://dwm.suckless.org/patches/attachaside/][attachaside]]
** [[https://dwm.suckless.org/patches/cfacts/][cfacts]]
** [[https://dwm.suckless.org/patches/restartsig/][restartsig]]
** [[https://dwm.suckless.org/patches/stacker/][stacker]]

* Volume Shortcuts
#include <X11/XF86keysym.h>


// ...

static const char *volup[]   = { "amixer", "-q", "set", "Master", "5%+", "unmute", NULL };
static const char *voldown[] = { "amixer", "-q", "set", "Master", "5%-", "unmute", NULL };
static const char *volmute[] = { "amixer", "-q", "set", "Master", "toggle",      NULL };

// ...

{ MODKEY,                       XK_F1,     spawn,          {.v = volmute } },
{ MODKEY,                       XK_F2,     spawn,          {.v = voldown } },
{ MODKEY,                       XK_F3,     spawn,          {.v = volup   } },



.emacs.d/config.el
(setq inhibit-startup-message t) ;; Disable Message

(scroll-bar-mode -1) ;;Disable visible scrollbar

(menu-bar-mode -1) ;;Disable the menu bar

(setq visible-bell -1) ;;Set up the visible bell

(show-paren-mode 1) ;; Enable ShowParent

(tool-bar-mode -1) ;;Disable the toolbar

(tooltip-mode -1) ;;Disable tooltips

(column-number-mode) ;;:Enable number
(global-display-line-numbers-mode t)

(dolist (mode '(org-mode-hook
                term-mode-hook
                shell-mode-hook
                eshell-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 0))))

(setq browse-url-generic-program 
      (executable-find "firefox-bin")
      browse-url-browser-function 'browse-url-generic)

(global-set-key (kbd "<escape>") 'keyboard-escape-quit)

;;(set-face-attribute 'default nil :font "Fira Code Retina" :height 101)

;; Set the fixed pitch face
;;(set-face-attribute 'fixed-pitch nil :font "Fira Code Retina" :height 101)

;; Set the variable pitch face
;;(set-face-attribute 'variable-pitch nil :font "Cantarell" :height 101 :weight 'regular)

(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("org" . "https://orgmode.org/elpa/")
                         ("elpa" . "https://elpa.gnu.org/packages/")))
(package-initialize)    
(unless package-archive-contents
  (package-refresh-contents))

(require 'use-package)
(setq use-package-always-ensure t)

(use-package auto-package-update
  :custom
  (auto-package-update-interval 7)
  (auto-package-update-prompt-before-update t)
  (auto-package-update-hide-results t)
  :config
  (auto-package-update-maybe)
  (auto-package-update-at-time "09:00"))

(use-package doom-themes
    :init
    (load-theme 'doom-nord t))

(use-package doom-modeline
  :init (doom-modeline-mode 1)
  :custom (doom-modeline-height 15))

(use-package all-the-icons)

(use-package all-the-icons-ivy-rich
  :defer 1
  :config(all-the-icons-ivy-rich-mode 1))

(use-package all-the-icons-dired
  :hook (dired-mode . all-the-icons-dired-mode))

(defun dw/org-mode-setup ()
  (org-indent-mode)
  ;; (variable-pitch-mode 1)
  (auto-fill-mode 0)
  (visual-line-mode 1)
  (setq evil-auto-indent nil))

(use-package org
  :hook (org-mode . dw/org-mode-setup)
  :config
  (setq org-ellipsis " ▾"
        org-hide-emphasis-markers t))


(use-package org-superstar
  :after org
  :hook (org-mode . org-superstar-mode)
  :custom
  (org-superstar-remove-leading-stars t)
  (org-superstar-headline-bullets-list '("◉" "○" "●" "○" "●" "○" "●")))



(with-eval-after-load 'org-faces
  (set-face-attribute 'org-document-title nil :font "Cantarell" :weight 'bold :height 1.3)
  (dolist (face '((org-level-1 . 1.2)
                  (org-level-2 . 1.1)
                  (org-level-3 . 1.05)
                  (org-level-4 . 1.0)
                  (org-level-5 . 1.1)
                  (org-level-6 . 1.1)
                  (org-level-7 . 1.1)
                  (org-level-8 . 1.1)))
    (set-face-attribute (car face) nil :font "Cantarell" :weight 'regular :height (cdr face)))


  (set-face-attribute 'org-block nil    :foreground nil :inherit 'fixed-pitch)
  (set-face-attribute 'org-table nil    :inherit 'fixed-pitch)
  (set-face-attribute 'org-formula nil  :inherit 'fixed-pitch)
  (set-face-attribute 'org-code nil     :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-table nil    :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-checkbox nil  :inherit 'fixed-pitch))


(defun dw/org-mode-visual-fill ()
  (setq visual-fill-column-width 100
        visual-fill-column-center-text t)
  (visual-fill-column-mode 1))

(use-package visual-fill-column
  :hook (org-mode . dw/org-mode-visual-fill))

(use-package ivy
  :defer 5
  :diminish ivy-mode
  :bind(("C-s" . swiper)
        :map ivy-minibuffer-map
        ("TAB" . ivy-alt-done)
        ("C-f" . ivy-alt-done)
        ("C-l" . ivy-alt-done)
        ("C-j" . ivy-next-line)
        ("C-k" . ivy-previous-line)
        :map ivy-switch-buffer-map
        ("C-k" . ivy-previous-line)
        ("C-l" . ivy-done)
        ("C-d" . ivy-switch-buffer-kill)
        :map ivy-reverse-i-search-map
        ("C-k" . ivy-previous-line)
        ("C-d" . ivy-reverse-i-search-kill))
  :config
  (ivy-mode 1))

(use-package which-key ;;
  :defer 5
  :diminish which-key-mode
  :config (which-key-mode
           (setq which-key-idle-delay 2)))

(use-package ivy-rich
  :defer 7
  :config
  (ivy-rich-mode 1))

(use-package counsel
  :bind (("M-x" . counsel-M-x)
         ("C-x b" . counsel-ibuffer)
         ("C-x C-f" . counsel-find-file)
         :map minibuffer-local-map
         ("C-r" . 'counsel-minibuffer-history))
  :config
  (setq ivy-initial-inputs-alist nil)) ;; Don't start searches with ^

(use-package web-mode
  :config
  (add-to-list 'auto-mode-alist '("\\.html?\\'" . web-mode))
  (progn
    (setq web-mode-engines-alist
          '(("django"  . "\\.html\\'")))
    (setq web-mode-ac-sources-alist
          '(("css" . (ac-source-css-property))
            ("html" . (ac-source-work-in-buffer ac-source-abbrev))))
    (setq web-mode-enable-auto-closing t)
    (setq web-mode-enable-auto-quoting t)
    (setq web-mode-enable-current-column-highlight t)
    (setq web-mode-enable-current-element-highlight t)))

(use-package emmet-mode
  :defer 5
  :hook
  (sgml-mode . emmet-mode) ;; Auto-start on any markup modes
  (web-mode . emmet-mode) ;; Auto-start on any markup modes
  (css-mode . emmet-mode)) ;; enable Emmet's css abbreviation.

(use-package company
  :defer 5
  :hook
  (after-init . global-company-mode))

(use-package skewer-mode
  :defer 5
  :hook
  (css-mode .skewer-css-mode)
  (html-mode . skewer-html-mode))

(use-package simple-httpd
  :defer t)

(use-package rjsx-mode
	  :mode "\\.js\\'")

(defun setup-tide-mode ()
  (interactive)
  (tide-setup)
  (flycheck-mode +1)
  (setq flycheck-check-syntax-automatically '(save mode-enabled))
  (eldoc-mode +1)
  (tide-hl-identifier-mode +1)
  (company-mode +1))
(setq company-tooltip-align-annotations t)
(add-hook 'before-save-hook 'tide-format-before-save)
(add-hook 'typescript-mode-hook #'setup-tide-mode)

(use-package tide 
  :defer 5
  :after (rjsx-mode company flycheck)
  :hook (rjsx-mode . setup-tide-mode))

(use-package prettier-js
  :defer 5
  :after (rjsx-mode)
  :hook (rjsx-mode . prettier-js-mode))

(use-package flycheck
  :defer 5
  :config
  (global-flycheck-mode))
(setq-default flycheck-disabled-checkers '(emacs-lisp-checkdoc))

(use-package magit
  :commands (magit-status magit-get-current-branch)
  :custom
  (magit-display-buffer-function #'magit-display-buffer-same-window-except-diff-v1))

(use-package projectile
  :diminish projectile-mode
  :config (projectile-mode)
  :demand t
  :bind ("C-M-p" . projectile-find-file)
  :bind-keymap
  ("C-c p" . projectile-command-map)
  :init
  (when (file-directory-p "~/Projects/Code")
    (setq projectile-project-search-path '("~/Projects/Code")))
  (setq projectile-switch-project-action #'projectile-dired))

(defun dw/after-tracking-remove-buffer (buffer)
  (dw/update-polybar-telegram))

(advice-add 'tracking-add-buffer :around #'dw/around-tracking-add-buffer)
(advice-add 'tracking-remove-buffer :after #'dw/after-tracking-remove-buffer)
(advice-remove 'tracking-remove-buffer #'dw/around-tracking-remove-buffer)

;; Advise exwm-workspace-switch so that we can more reliably clear tracking buffers
;; NOTE: This is a hack and I hate it.  It'd be great to find a better solution.
(defun dw/before-exwm-workspace-switch (frame-or-index &optional force)
  (when (fboundp 'tracking-remove-visible-buffers)
    (when (eq exwm-workspace-current-index 0)
      (tracking-remove-visible-buffers))))

(advice-add 'exwm-workspace-switch :before #'dw/before-exwm-workspace-switch)

(use-package telega
  :commands telega
  :config
  (setq telega-user-use-avatars nil
        telega-use-tracking-for '(any pin unread)
        telega-chat-use-markdown-formatting t
        telega-emoji-use-images t
        telega-completing-read-function #'ivy-completing-read
        telega-msg-rainbow-title nil
        telega-chat-fill-column 75))

(use-package dashboard
  :ensure t
  :init
  (progn
    (setq dashboard-items '((projects . 5) 
                            (recents . 10))) 
    (setq dashboard-show-shortcuts nil)
    (setq dashboard-center-content nil)
     ;;(setq dashboard-banner-logo-title "Lindo, siono raza?")
    (setq dashboard-set-file-icons t)
    (setq dashboard-set-heading-icons t)
    ;;(setq dashboard-startup-banner "~/.emacs.d/logo/logo.png")
    (setq dashboard-set-navigator t)
    )
  :config
  (dashboard-setup-startup-hook))

(setq dashboard-navigator-buttons
      `(;; line1
        ((,(all-the-icons-octicon "mark-github" :height 1.1 :v-adjust 0.0)
          "Homepage"
          "Browse homepage"
          (lambda (&rest _) (browse-url "https://github.com/a-bbi")))
         ("★" nil "Show stars" (lambda (&rest _) (show-stars)) warning))
        ;; line 2
        ;; ((,(all-the-icons-faicon "linkedin" :height 1.1 :v-adjust 0.0)
        ;;   "Linkedin"
        ;;   ""
        ;;   (lambda (&rest _) (browse-url "homepage")))
        ;;  ("⚑" nil "Show flags" (lambda (&rest _) (message "flag")) error))
        ))

(use-package ace-window
  :bind
  ("C-x o" . ace-window)
  :config
  (setq aw-keys '(?a ?s ?d ?f)))

(use-package rainbow-delimiters
  :defer 5
  :hook (prog-mode . rainbow-delimiters-mode))

(use-package command-log-mode
  :defer t)

(use-package helpful
  :defer 5
  :custom
  (counsel-describe-function-function #'helpful-callable)
  (counsel-describe-variable-function #'helpful-variable)
  :bind
  ([remap describe-function] . helpful-function)
  ([remap describe-symbol] . helpful-symbol)
  ([remap describe-variable] . helpful-variable)
  ([remap describe-command] . helpful-command)
  ([remap describe-key] . helpful-key))
